### PR TITLE
[utils] implement Verhoeff checksum calculation and validation

### PR DESCRIPTION
**Pull request by abtink**
This commit adds `Utils::VerhoeffChecksum` class along with public OT APIs and related CLI commands for Verhoeff checksum calculation and validation. Unit test `test_checksum` is updated to test the new module.

---

Related to [SPEC-1216](https://threadgroup.atlassian.net/browse/SPEC-1216)

### CODE REVIEW

File: src/core/config/misc.h [hunk 1]
```diff
@@ -141,6 +141,16 @@
 #define OPENTHREAD_CONFIG_JAM_DETECTION_ENABLE 0
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_VERHOEFF_CHECKSUM_ENABLE
+ *
+ * Define to 1 toe enable Verhoeff checksum utility module.
```
Comment (EskoDijk) [inline src/core/config/misc.h:147]:
```suggestion
 * Define to 1 to enable Verhoeff checksum utility module.
```

**Comment by EskoDijk** [comment]
Just wondering why the size-report bot reports a large size increase for ot-cli-ftd: since by default this code is not included in the build? (only on border-agents)   I would assume the actual size increase would be +0 for non-BAs.

Or perhaps it's better to only enable this feature in the build, if the "ePSKc" feature is also enabled. This would be the case for newer Thread BRs only.

File: src/core/config/misc.h (hunk 1 above)
Comment (abtink) [reply src/core/config/misc.h:147]: Thanks.

**Comment by abtink** [comment]
> Just wondering why the size-report bot reports a large size increase for ot-cli-ftd: since by default this code is not included in the build? (only on border-agents) I would assume the actual size increase would be +0 for non-BAs.

Yes. The `OPENTHREAD_CONFIG_BORDER_AGENT_ENABLE` is enabled in `ot-core-config-check-size-ftd.h` (may make sense to remove it later)?

> Or perhaps it's better to only enable this feature in the build, if the "ePSKc" feature is also enabled. This would be the case for newer Thread BRs only.

Make sense to do so (will change it in the next push).
But I think it will not impact the code-size report since `OPENTHREAD_CONFIG_BORDER_AGENT_EPHEMERAL_KEY_ENABLE` is also enabled in `ot-core-config-check-size-ftd.h.`.

**Comment by abtink** [comment]
@EskoDijk one thing puzzles me, the [wikipedia article](https://en.wikipedia.org/wiki/Verhoeff_algorithm) works through one example in the algorithm description:

> For example the check digit for 248 is 5.

But "2485" does not seem to pass as valid. This seems to be the case testing it with other available implementations of Verhoeff.

**Comment by Vyrastas** [comment]
@jrhodie can you take a look at this one too please

File: src/cli/cli.cpp [hunk 2]
```diff
@@ -8355,6 +8356,57 @@ void Interpreter::HandleIp6Receive(otMessage *aMessage, void *aContext)
...
 #endif
 
+#if OPENTHREAD_CONFIG_VERHOEFF_CHECKSUM_ENABLE
+
+template <> otError Interpreter::Process<Cmd("verhoeff")>(Arg aArgs[])
+{
+    otError error;
+
+    /**
+     * @cli verhoeff calculate <decimalstring>
```
Comment (jrhodie) [inline src/cli/cli.cpp:8366]: can remove <decimalstring> from the @cli tag, as it doesn't show up in the output anyway. But please include parameters with the @cparam tag.

File: src/cli/cli.cpp [hunk 3]
```diff
@@ -8355,6 +8356,57 @@ void Interpreter::HandleIp6Receive(otMessage *aMessage, void *aContext)
...
+    if (aArgs[0] == "calculate")
+    {
+        char checksum;
+
+        VerifyOrExit(!aArgs[1].IsEmpty() && aArgs[2].IsEmpty(), error = OT_ERROR_INVALID_ARGS);
+        SuccessOrExit(error = otVerhoeffChecksumCalculate(aArgs[1].GetCString(), &checksum));
+        OutputLine("%c", checksum);
+    }
+    /**
+     * @cli verhoeff validate <decimalstring>
```
Comment (jrhodie) [inline src/cli/cli.cpp:8384]: can remove from the @cli tag, as it doesn't show up in the output anyway. But please include parameters with the @cparam tag.

File: src/cli/cli.cpp [hunk 4]
```diff
@@ -8355,6 +8356,57 @@ void Interpreter::HandleIp6Receive(otMessage *aMessage, void *aContext)
...
+     * @code
+     * verhoeff validate 307318421
+     * Done
+     * verhoeff validate 307318425
```
Comment (jrhodie) [inline src/cli/cli.cpp:8388]: Perhaps include some explanation here - are two different strings being validated?

**Review by jrhodie — COMMENTED** [review]
just a couple of comments on the commands. 

 Also, I will do a CL to add OPENTHREAD_CONFIG_VERHOEFF_CHECKSUM_ENABLE  to our Doxyfile local file so that the new Verhoeff commands appear in the Command Ref.

File: src/cli/cli.cpp (hunk 2 above)
Comment (abtink) [reply src/cli/cli.cpp]: Updates in new push. Thanks.

File: src/cli/cli.cpp (hunk 3 above)
Comment (abtink) [reply src/cli/cli.cpp]: Updates in new push. Thanks.

File: src/cli/cli.cpp (hunk 4 above)
Comment (abtink) [reply src/cli/cli.cpp]: I decided to remove the second case (make it simpler).

**Review by jrhodie — APPROVED** [review]

**Review by EskoDijk — APPROVED** [review]
Thanks, looks good - tested the CLI with some values and it validates/fails as expected.

**Review by Vyrastas — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**